### ISSUE TITLE
How do I add default constructors to the entity classes?

### ISSUE THREAD

**Issue by glenlewisdd** [ErikEJ/EFCorePowerTools#1378]
I would like to have the default constructor be like this if possible

If I can modify a template or something simple I will do it.  I use the context often for the partial classes and EF 6 Core doesn't have an easy way to access the entity context like I was able to in EF6.   If I can add the constructor with the context simply then EF 6 will DI the parameters as needed.  

```
public partial class MyDbContext 
{
    protected MyDbContext Context;
    public MyEntity(MyDbContext context)
    {
        Context = context;
    }
    
    /****
    Properties Code.....
    ***/
}
```

**Comment by ErikEJ** [comment]
Maybe the Handelbars templates?

**Comment by glenlewisdd** [comment]
Do you have any examples here I can look at to implement the handelbars templates.  I have never worked with them before.

Thank you.

**Comment by ErikEJ** [comment]
There are some linked from the wiki.

**Comment by glenlewisdd** [comment]
Last question on the handelbars template.

Where can I find the variables so I can map it correctly.
For example:

My DBContext name is VPEntities how can I map the class name of the DBContext so it isn't hardcoded within the template?

```
{{> imports}}

namespace {{namespace}}
{
    {{#if comment}}
    /// <summary>
{{comment}}
    /// </summary>
    {{/if}}
    {{#each class-annotations}}
    {{{class-annotation}}}
    {{/each}}
    public partial class {{class}}
    {
        public VPEntities DB;
        public {{class}}(VPEntities _db)
        {
            DB = _db;
        }

        {{{> constructor}}}
        {{{> properties}}}
    }
}
```

**Comment by ErikEJ** [comment]
Not sure it is possible. @tonysneed ?

**Comment by tonysneed** [comment]
I believe you can do this with the [Custom Template Data](https://github.com/TrackableEntities/EntityFrameworkCore.Scaffolding.Handlebars#custom-template-data) feature. This lets you specify the name of your DbContext class as a value in the `options.TemplateData` dictionary when you call `services.AddHandlebarsScaffolding`.

**Comment by ErikEJ** [comment]
Thanks. So not in Power Tools.

**Comment by tonysneed** [comment]
@glenlewisdd The pattern of injecting a DBContext into entities looks suspect to me. What is it you are trying to accomplish?

**Comment by 2xRon** [comment]
For us, it's a big (EF6) code base that relies on proxy-entities carrying around their context so child controls can use it to load other entities on the same context (e.g., foreign key pick lists). In EF6 you could do this with reflection, but I don't think that still works in EF6Core.

**Comment by ErikEJ** [comment]
I do not think there is any compelling reason to do this, lazy loading is possible with EF Core using other means, and you can always create partial classes with the constructor as needed.

**Comment by glenlewisdd** [comment]
@tonysneed the database I am connect to is apart of an ERP system.  I have setup as much relationships as possible to use the navigational properties.  But what I have found and maybe this was resolved in EF 6 Core, is when you navigate down and then back up to a parent object to navigate to a new relationship.  EF pulls in the entire table and not just filter to what you have linq where statment on.

Example.

TAbles:
Companies -> Jobs
Companies -> Customer

If I grab a job then navigate back to company then to the Customer to get info.  EF 6 would return the entire customer table then filter to the customer.  

```
var job = context.JObs.Where(f => f.JobId == 12345);

//This navigation results in a full table pull then filter in memory
var cust = job.Company.Customers.FirstOrDefault(f=>f.CustomerId == 123);

//So rather I used this, this result is only returning from the Database the single record.
var cust = job.DB.Customers.FirstOrDefault(f => f.CustomerCompanyId == job.CustomerCompanyId  && f.CustomerId == 123);

```
This is just a basic example, not the entire reason for having the context available.  
Another reason is to attach an existing object from the same context to a new entity.  (If you attach an object from a different Context EF will think it is a new entity and try to add it, causing an error in the save operation).

@ErikEJ  The reason to add it to the template is to ensure I always have access to the context if needed.  Adding it to the partial results in the same outcome but requires me to add it to each entity manually.  (I am lazy and if I can have templates kick out my code I will :-) ).

I have just found from a performance stand point getting the data directly from the context can be faster than the EF query it generates for lazy loading of the navigation.  (Again if you traverse to a parent entity and back down to a different child it pull the entire table then filters).

Thank you all for the input, @tonysneed  I will look into using the Custom Template data to create the variable.
